Cython/Utility/Dataclasses.c

```diff
@@ -0,0 +1,78 @@
+///////////////////// ModuleLoader.proto //////////////////////////
+
+static PyObject* __Pyx_LoadInternalModule(const char* name, const char* fallback_code); /* proto */
+
+//////////////////// ModuleLoader ///////////////////////
+//@requires: CommonStructures.c::FetchSharedCythonModule
+
+static PyObject* __Pyx_LoadInternalModule(const char* name, const char* fallback_code) {
+    // In supporting dataclasses we want to be able to use directives the contents
+    // of the standard library dataclasses module where available
+    // If those objects aren't available (due to Python version) then a simple fallback is substituted
+    // instead, which largely just fails with a not-implemented error.
+    //
+    // The fallbacks are placed in the "shared abi module" as a convenient internal place to
+    // store them
+
+    PyObject *shared_abi_module = 0, *module = 0;
+
+    shared_abi_module = __Pyx_FetchSharedCythonABIModule();
+    if (!shared_abi_module) return NULL;
+
+    if (PyObject_HasAttrString(shared_abi_module, name)) {
+        PyObject* result = PyObject_GetAttrString(shared_abi_module, name);
+        Py_DECREF(shared_abi_module);
+        return result;
+    }
+
+    // the best and simplest case is simply to defer to the standard library (if available)
+    module = PyImport_ImportModule(name);
+    if (!module) {
+        PyObject *localDict, *runValue, *builtins, *modulename;
+        if (PyErr_ExceptionMatches(PyExc_ImportError)) {
+            PyErr_Clear();  // this is reasonably likely (especially on older versions of Python)
+        } else {
+            goto bad;
+        }
+        modulename = PyBytes_FromFormat("_cython_" CYTHON_ABI ".%s", name);
+        if (!modulename) goto bad;
+#if PY_MAJOR_VERSION >= 3 && CYTHON_COMPILING_IN_CPYTHON
+        module = PyImport_AddModuleObject(modulename); // borrowed
+#else
+        module = PyImport_AddModule(PyBytes_AsString(modulename)); // borrowed
+#endif
+        Py_DECREF(modulename);
+        if (!module) goto bad;
+        Py_INCREF(module);
```

Comment:
Isn't this duplicated code? We probably need something like this wherever we use the ABI module, right?

Reply:
I don't think so... The main difference here is that I provide a small amount of Python fallback code so that the entries in `__dataclass_fields__` always have a type. I don't think the ABI module provides something to do that.